### PR TITLE
#373 - Kong subscription

### PR DESCRIPTION
**Pull request by vladyslav-fenchak**

### CODE REVIEW

File: packages/core/minos-microservice-networks/minos/networks/decorators/definitions/abc.py [hunk 1]
```diff
@@ -30,6 +30,9 @@ class EnrouteDecorator(ABC):
     # noinspection PyFinal
     KIND: Final[EnrouteDecoratorKind]
 
+    def __init__(self, **kwargs):
+        self.kwargs = kwargs
```
Comment (garciparedes) [inline packages/core/minos-microservice-networks/minos/networks/decorators/definitions/abc.py:34]: Could you add a test to ensure that additional named arguments will be stored on `kwargs` variable?

File: packages/core/minos-microservice-networks/minos/networks/decorators/definitions/http/abc.py [hunk 2]
```diff
@@ -12,7 +12,8 @@
 class HttpEnrouteDecorator(EnrouteDecorator):
     """Http Enroute Decorator class."""
 
-    def __init__(self, path: Optional[str] = None, method: Optional[str] = None, url: Optional[str] = None):
+    def __init__(self, path: Optional[str] = None, method: Optional[str] = None, url: Optional[str] = None, **kwargs):
+        super().__init__(**kwargs)
```
Comment (garciparedes) [inline packages/core/minos-microservice-networks/minos/networks/decorators/definitions/http/abc.py:16]: Could you add `kwargs` propagation to the remaining `EnrouteDecorator` heirs, like `BrokerEnrouteDecorator`, `PeriodicEnrouteDecorator`, `GraphQlEnrouteDecorator` and so on?

File: packages/plugins/minos-router-graphql/tests/utils.py [hunk 3]
```diff
@@ -79,6 +79,26 @@ def get_order(self, request: Request):
...
+    # noinspection PyUnusedLocal
+    @enroute.graphql.query(
+        name="order",
+        argument=GraphQLField(GraphQLString),
+        output=GraphQLString,
+        authorized=True,
+        allowed_groups=[
+            "supèr_admin",
+            "admin",
+        ],
```
Comment (garciparedes) [inline packages/plugins/minos-router-graphql/tests/utils.py:94]: I think a simple `foo="bar"` is enough to check `kwargs`.

File: packages/core/minos-microservice-networks/tests/test_networks/test_decorators/test_api.py [hunk 4]
```diff
@@ -16,10 +16,56 @@ def test_rest_command(self):
...
 
+    def test_rest_command_kwargs(self):
+        decorator = enroute.rest.command(
+            path="tickets/",
+            method="GET",
+            authorized=True,
+            allowed_groups=[
+                "super_admin",
+                "admin",
+            ],
```
Comment (garciparedes) [inline packages/core/minos-microservice-networks/tests/test_networks/test_decorators/test_api.py:27]: I think a simple `foo="bar"` is enough to check `kwargs`.

File: packages/core/minos-microservice-networks/tests/test_networks/test_decorators/test_api.py [hunk 5]
```diff
@@ -16,10 +16,56 @@ def test_rest_command(self):
...
 
+    def test_rest_query_kwargs(self):
+        decorator = enroute.rest.query(
+            path="tickets/",
+            method="GET",
+            authorized=True,
+            allowed_groups=[
+                "super_admin",
+                "admin",
+            ],
```
Comment (garciparedes) [inline packages/core/minos-microservice-networks/tests/test_networks/test_decorators/test_api.py:54]: I think a simple `foo="bar"` is enough to check `kwargs`.

File: packages/core/minos-microservice-networks/minos/networks/discovery/connectors.py [hunk 6]
```diff
@@ -104,7 +101,8 @@ def _endpoints_from_config(config: Config) -> list[dict[str, Any]]:
         for name in config.get_services():
             decorators = EnrouteCollector(name, config).get_rest_command_query()
             endpoints += [
-                {"url": decorator.url, "method": decorator.method} for decorator in set(chain(*decorators.values()))
+                {"url": decorator.url, "method": decorator.method} | decorator.kwargs
+                for decorator in set(chain(*decorators.values()))
```
Comment (garciparedes) [inline packages/core/minos-microservice-networks/minos/networks/discovery/connectors.py:105]: Could you add a test to check this code?

File: packages/core/minos-microservice-networks/minos/networks/discovery/connectors.py [hunk 7]
```diff
@@ -79,10 +79,7 @@ def _client_from_config(cls, config: Config) -> DiscoveryClient:
         discovery_config = config.get_discovery()
 
         client_cls = cls._client_cls_from_config(discovery_config)
-        client_host = discovery_config.get("host")
-        client_port = discovery_config.get("port")
-
-        return client_cls(host=client_host, port=client_port)
+        return client_cls.from_config(config)
```
Comment (garciparedes) [inline packages/core/minos-microservice-networks/minos/networks/discovery/connectors.py:82]: Could you add a test to check this code?

File: packages/core/minos-microservice-networks/minos/networks/discovery/clients/abc.py [hunk 8]
```diff
@@ -28,6 +33,14 @@ def route(self) -> str:
         # noinspection HttpUrlsUsage
         return f"http://{self.host}:{self.port}"
 
+    @classmethod
+    def _from_config(cls, config: Config, **kwargs) -> DiscoveryClient:
+        discovery_config = config.get_discovery()
+
+        client_host = discovery_config.get("host")
+        client_port = discovery_config.get("port")
+        return cls(host=client_host, port=client_port, **kwargs)
```
Comment (garciparedes) [inline packages/core/minos-microservice-networks/minos/networks/discovery/clients/abc.py:42]: Could you add a test to check this code?

File: packages/plugins/minos-discovery-kong/tests/test_kong/test_middleware.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+
+async def _fn(request: Request) -> Response:
+    return Response(f"{await request.content()}bar")
+
+
+Raw = namedtuple("Raw", ["headers", "content"])
+
+
+class TestKongDiscoveryClient(unittest.IsolatedAsyncioTestCase):
```
Comment (garciparedes) [inline packages/plugins/minos-discovery-kong/tests/test_kong/test_middleware.py:30]: Could you add a test that checks the case in which `"X-Consumer-Custom-ID"` is not provided?

File: packages/plugins/minos-http-aiohttp/minos/plugins/aiohttp/requests.py [hunk 10]
```diff
@@ -90,14 +90,14 @@ def user(self) -> Optional[UUID]:
...
 
-    @property
+    @cached_property
     def headers(self) -> dict[str, str]:
         """Get the headers of the request.
 
         :return: A dictionary in which keys are ``str`` instances and values are ``str`` instances.
         """
         # noinspection PyTypeChecker
-        return self.raw.headers
```
Comment (garciparedes) [inline packages/plugins/minos-http-aiohttp/minos/plugins/aiohttp/requests.py:100]: Could you add a test that checks if `headers` attribute can be modified?

File: packages/plugins/minos-discovery-kong/minos/plugins/kong/kong_client.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
+from uuid import (
```
Comment (garciparedes) [inline packages/plugins/minos-discovery-kong/minos/plugins/kong/kong_client.py:1]: Instead of naming this file as `kong_client.py` I think you can simply name it as `client.py`. To avoid a collision with the one that contains de `KongDiscoveryClient` class, you can rename it as `discovery.py`

File: packages/plugins/minos-discovery-kong/README.md [hunk 12]
```diff
@@ -27,7 +27,6 @@ Modify `config.yml` file:
 ```yaml
 ...
 discovery:
-  connector: minos.networks.DiscoveryConnector
```
Comment (garciparedes) [inline packages/plugins/minos-discovery-kong/README.md:30]: Why did you remove this line? I think with `minos.common.ConfigV2` is a mandatory field.

**Review by garciparedes — CHANGES_REQUESTED** [review]
Hi @vladyslav-fenchak, this PR look really great! However there are some comments I left around the code, could you review them?

File: packages/plugins/minos-router-graphql/tests/utils.py (hunk 3 above)
Comment (vladyslav-fenchak) [reply packages/plugins/minos-router-graphql/tests/utils.py]: Done!

File: packages/core/minos-microservice-networks/tests/test_networks/test_decorators/test_api.py (hunk 4 above)
Comment (vladyslav-fenchak) [reply packages/core/minos-microservice-networks/tests/test_networks/test_decorators/test_api.py:27]: Done!

File: packages/core/minos-microservice-networks/tests/test_networks/test_decorators/test_api.py (hunk 5 above)
Comment (vladyslav-fenchak) [reply packages/core/minos-microservice-networks/tests/test_networks/test_decorators/test_api.py:54]: Done!

File: packages/plugins/minos-discovery-kong/README.md (hunk 12 above)
Comment (vladyslav-fenchak) [reply packages/plugins/minos-discovery-kong/README.md:30]: Done!

File: packages/plugins/minos-discovery-kong/minos/plugins/kong/kong_client.py (hunk 11 above)
Comment (vladyslav-fenchak) [reply packages/plugins/minos-discovery-kong/minos/plugins/kong/kong_client.py:1]: Done!

File: packages/plugins/minos-discovery-kong/tests/test_kong/test_middleware.py (hunk 9 above)
Comment (vladyslav-fenchak) [reply packages/plugins/minos-discovery-kong/tests/test_kong/test_middleware.py:30]: Done!

File: packages/plugins/minos-http-aiohttp/minos/plugins/aiohttp/requests.py (hunk 10 above)
Comment (vladyslav-fenchak) [reply packages/plugins/minos-http-aiohttp/minos/plugins/aiohttp/requests.py:100]: Done!

File: packages/core/minos-microservice-networks/minos/networks/discovery/connectors.py (hunk 6 above)
Comment (vladyslav-fenchak) [reply packages/core/minos-microservice-networks/minos/networks/discovery/connectors.py:105]: Done!

File: packages/core/minos-microservice-networks/minos/networks/discovery/connectors.py (hunk 7 above)
Comment (vladyslav-fenchak) [reply packages/core/minos-microservice-networks/minos/networks/discovery/connectors.py:82]: Done!

File: packages/core/minos-microservice-networks/minos/networks/discovery/clients/abc.py (hunk 8 above)
Comment (vladyslav-fenchak) [reply packages/core/minos-microservice-networks/minos/networks/discovery/clients/abc.py:42]: Done!

File: packages/plugins/minos-discovery-kong/minos/plugins/kong/__init__.py [hunk 13]
```diff
@@ -2,6 +2,10 @@
 __email__ = "hey@minos.run"
 __version__ = "0.1.0a1"
 
-from .client import (
+from .discovery import (
+    KongClient,
     KongDiscoveryClient,
 )
```
Comment (garciparedes) [inline packages/plugins/minos-discovery-kong/minos/plugins/kong/__init__.py:8]:
```suggestion
from .client import (
    KongClient,
)
from .discovery import (
    KongDiscoveryClient,
)
```

File: packages/plugins/minos-router-graphql/tests/utils.py [hunk 14]
```diff
@@ -79,6 +79,26 @@ def get_order(self, request: Request):
...
+    # noinspection PyUnusedLocal
+    @enroute.graphql.query(
+        name="order",
+        argument=GraphQLField(GraphQLString),
+        output=GraphQLString,
+        authorized=True,
+        allowed_groups=[
+            "foo",
+            "bar",
+        ],
```
Comment (garciparedes) [inline packages/plugins/minos-router-graphql/tests/utils.py:94]:
```suggestion
        foo="bar",
```

File: packages/plugins/minos-router-graphql/tests/test_graphql/test_decorators.py [hunk 15]
```diff
@@ -36,6 +37,27 @@ def test_get_all_queries(self):
...
+            "get_order": {
+                GraphQlQueryEnrouteDecorator(
+                    name="order",
+                    argument=GraphQLField(GraphQLString),
+                    output=GraphQLString,
+                    authorized=True,
+                    allowed_groups=[
+                        "foo",
+                        "bar",
+                    ],
```
Comment (garciparedes) [inline packages/plugins/minos-router-graphql/tests/test_graphql/test_decorators.py:54]:
```suggestion
                    foo="bar",
```

**Review by garciparedes — APPROVED** [review]
Great PR @vladyslav-fenchak!
